bitwarden_access_manager.py

```diff
@@ -0,0 +1,737 @@
+                    for x in data["member_access"]
+                )
+            )
+
+        return Collection(
+            id=data["collection_id"],
+            external_id=data["external_id"],
+            group_access=group_access,
+            member_access=member_access,
+        )
+
+    def format_toml(self) -> str:
+        result = (
+            "[[collection]]\n"
+            f'collection_id = "{self.id}"\n'
+            f'external_id = "{self.external_id}"\n'
+        )
+
+        if self.member_access is not None:
+            member_access_lines = [
+                "  " + a.format_toml() for a in sorted(self.member_access)
+            ]
+            if len(member_access_lines) > 0:
+                result = (
+                    result
+                    + "member_access = [\n"
+                    + ",\n".join(member_access_lines)
+                    + ",\n]\n"
+                )
+            else:
+                result = result + "member_access = []\n"
+
+        if self.group_access is not None:
+            group_access_lines = [
+                "  " + a.format_toml() for a in sorted(self.group_access)
+            ]
+            if len(group_access_lines) > 0:
+                result = (
+                    result
+                    + "group_access = [\n"
+                    + ",\n".join(group_access_lines)
+                    + ",\n]"
+                )
+            else:
+                result = result + "group_access = []"
+        return result
+
+
+class BitwardenClient(NamedTuple):
+    connection: HTTPSConnection
+    bearer_token: str
+
+    @staticmethod
+    def new(client_id: str, client_secret: str) -> BitwardenClient:
+        response = requests.post(
+            "https://identity.bitwarden.com/connect/token",
+            headers={"Content-Type": "application/x-www-form-urlencoded"},
+            data={
+                "grant_type": "client_credentials",
+                "scope": "api.organization",
+                "Accept": "application/json",
+            },
+            auth=(client_id, client_secret),
+        )
+        bearer_token = response.json()["access_token"]
+        return BitwardenClient(HTTPSConnection("api.bitwarden.com"), bearer_token)
+
+    def _http_get(self, url: str) -> HTTPResponse:
+        self.connection.request(
+            method="GET",
+            url=url,
+            headers={
+                "Accept": "application/json",
+                "Authorization": f"Bearer {self.bearer_token}",
+            },
+        )
+
+        return self.connection.getresponse()
+
+    def get_groups(self) -> Iterable[Group]:
+        groups = json.load(self._http_get(f"/public/groups"))
+        for group in groups["data"]:
+            yield Group(
+                id=group["id"],
+                name=group["name"],
+                access_all=group["accessAll"],
+            )
+
+    def get_collection_groups(self, groups: Any) -> Iterable[GroupCollectionAccess]:
```

Comment:
```suggestion
    def get_collection_groups(self, groups: Dict[str, Any]) -> Iterable[GroupCollectionAccess]:
```
We can be slightly more specific with the type.

Reply:
When I change to `Dict[str, Any]`:
```
   377      def get_collection_groups(
   378          self, groups: Dict[str, Any]
   379      ) -> Iterable[GroupCollectionAccess]:
   380          for group in groups:
   381              access = self.check_access(bool(group["readonly"]))
   382
   383              group_id = group["id"]
   384              yield GroupCollectionAccess(
   385                  name=json.load(self._http_get(f"/public/groups/{group_id}"))["name"],
   386                  access=access,
   387              )
```

Mypy is returning the following errors:
```
bitwarden_access_manager.py:381: error: Invalid index type "str" for "str"; expected type "Union[SupportsIndex, slice]"  [index]
bitwarden_access_manager.py:383: error: Invalid index type "str" for "str"; expected type "Union[SupportsIndex, slice]"  [index]
```

I don't understand what he's expecting here, I'm not using index to access the dict.

Reply:
As discussed off-channel, changing `groups: Dict[str, Any]` to `groups: List[Dict[str, Any]]` should resolve this.

Reply:
Just noticed that we don't need to fix this because the function has been removed in https://github.com/ChorusOne/github-access-manager/pull/10/commits/d097be834a50475e52a4b02e7894bbedde97aa5d

Thanks your the explanations !